patches/api/0363-Expand-FallingBlock-API.patch

```diff
@@ -0,0 +1,108 @@
++     * @deprecated Use {@link #spawn(Location, Class, Consumer)} (or a variation thereof) in combination with {@link FallingBlock#setBlockData(BlockData)}
+      */
+     @NotNull
++    @Deprecated // Paper
+     public FallingBlock spawnFallingBlock(@NotNull Location location, @NotNull MaterialData data) throws IllegalArgumentException;
+ 
+     /**
+@@ -2255,8 +2257,10 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+      * @return The spawned {@link FallingBlock} instance
+      * @throws IllegalArgumentException if {@link Location} or {@link
+      *     BlockData} are null
++     * @deprecated Use {@link #spawn(Location, Class, Consumer)} (or a variation thereof) in combination with {@link FallingBlock#setBlockData(BlockData)}
+      */
+     @NotNull
++    @org.jetbrains.annotations.ApiStatus.Obsolete // Paper
+     public FallingBlock spawnFallingBlock(@NotNull Location location, @NotNull BlockData data) throws IllegalArgumentException;
+ 
+     /**
+@@ -2273,7 +2277,7 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+      * @return The spawned {@link FallingBlock} instance
+      * @throws IllegalArgumentException if {@link Location} or {@link
+      *     Material} are null or {@link Material} is not a block
+-     * @deprecated Magic value
++     * @deprecated Magic value. Use {@link #spawn(Location, Class, Consumer)} (or a variation thereof) in combination with {@link FallingBlock#setBlockData(BlockData)}
+      */
+     @Deprecated
+     @NotNull
+diff --git a/src/main/java/org/bukkit/entity/FallingBlock.java b/src/main/java/org/bukkit/entity/FallingBlock.java
+index ae1ce2ee2deb82f3f4144ec54b3ba119b437c5e9..2db2992d09d2c67b8f9ec4c87d86a321cc288b0e 100644
+--- a/src/main/java/org/bukkit/entity/FallingBlock.java
++++ b/src/main/java/org/bukkit/entity/FallingBlock.java
+@@ -26,6 +26,33 @@ public interface FallingBlock extends Entity {
+      */
+     @NotNull
+     BlockData getBlockData();
++    // Paper start
++    /**
++     * Sets the data for the falling block.
++     * <br>
++     * The data on the stored {@link #getPlacedBlockState() placed BlockState} will be purged using this method.
++     *
++     * @param blockData the data to use as the block
++     */
++    void setBlockData(@NotNull BlockData blockData);
++
++    /**
++     * Get the data of the falling block represented as a {@link org.bukkit.block.BlockState BlockState}
++     * which includes potential NBT data that gets applied when the block gets placed on landing.
++     *
++     * @return the BlockState representing this block
++     */
++    @NotNull
++    org.bukkit.block.BlockState getPlacedBlockState();
++
++
++    /**
++     * Sets the {@link BlockData} and possibly present tile entity data for the falling block.
++     *
++     * @param blockState the BlockState to use
++     */
++    void setPlacedBlockState(@NotNull org.bukkit.block.BlockState blockState);
```

Comment:
Any reason for the "placed" prefix here ?

Reply:
I added that because the block state is not applied until the block is placed on landing.

But most blocks where that really matters aren't rendered properly, so it could be removed anyways

Reply:
Yea, I mean given the fact that this method immediately changes the "block data"/looks of the falling block, I think it would be best to drop the placed prefix.